File: src/rpc.js
import { createConnection } from "node:net";
import { statePaths } from "./state.js";

export function rpc(request, socketPath = statePaths().socket) {
  return new Promise((resolve, reject) => {
    const socket = createConnection(socketPath);
    let buffered = "";
    socket.on("connect", () => socket.write(`${JSON.stringify(request)}\n`));
    socket.on("data", chunk => {
      buffered += chunk;
      const newline = buffered.indexOf("\n");
      if (newline < 0) return;
      const response = JSON.parse(buffered.slice(0, newline));
      socket.end();
      if (response.ok) resolve(response.result);
      else reject(new Error(response.error));
    });
    socket.on("error", error => reject(new Error(`daemon unavailable: ${error.message}; start it with \`resync daemon\``)));
  });
}

File: src/rpc.rs
use crate::state::state_paths;
use anyhow::{Result, bail};
use serde_json::Value;
use std::io::{BufRead, BufReader, Write};
use std::os::unix::net::UnixStream;
use std::path::Path;

pub fn rpc(request: &Value, socket_path: Option<&Path>) -> Result<Value> {
    let path = socket_path
        .map(Path::to_owned)
        .unwrap_or_else(|| state_paths().socket);
    let mut stream = UnixStream::connect(&path).map_err(|error| {
        anyhow::anyhow!("daemon unavailable: {error}; start it with `resync daemon`")
    })?;
    serde_json::to_writer(&mut stream, request)?;
    stream.write_all(b"\n")?;
    let mut line = String::new();
    BufReader::new(stream).read_line(&mut line)?;
    let response: Value = serde_json::from_str(&line)?;
    if response.get("ok").and_then(Value::as_bool) == Some(true) {
        Ok(response.get("result").cloned().unwrap_or(Value::Null))
    } else {
        bail!(
            "{}",
            response
                .get("error")
                .and_then(Value::as_str)
                .unwrap_or("daemon request failed")
        )
    }
}
